File: ruby/testcode/massassign_033.rb
require_relative 'shared'

User = Struct.new(:name, :email, :role, :bio, :admin, keyword_init: true)
class << User; def create(h); new(**h.transform_keys(&:to_sym).slice(:name,:email,:role,:bio,:admin)); end; end

# vuln-code-snippet start ruby_massassign_dry_schema_filter
def handler(req)
  schema = { 'name' => String, 'email' => String }
  attrs = req.post_data.select { |k, v| schema.key?(k) && v.is_a?(schema[k]) }
  user = User.create(attrs) # vuln-code-snippet safe-line ruby_massassign_dry_schema_filter
  BenchmarkResponse.json({ ok: true })
end
# vuln-code-snippet end ruby_massassign_dry_schema_filter

File: php/testcode/massassign_033.php
<?php
require_once __DIR__ . '/shared.php';

// vuln-code-snippet start php_massassign_form_request_validated
function massassign033(BenchmarkRequest $req): BenchmarkResponse {
    $rules = ['name' => 'required|string', 'email' => 'required|email'];
    $validated = array_intersect_key($_POST, $rules); // vuln-code-snippet safe-line php_massassign_form_request_validated
    $model = new stdClass();
    foreach ($validated as $k => $v) {
        $model->$k = $v;
    }
    return BenchmarkResponse::ok('validated');
}
// vuln-code-snippet end php_massassign_form_request_validated
